jax_md/elasticity.py

```diff
@@ -0,0 +1,526 @@
+  a fourth-rank tensor of shape (dimension,dimension,dimension,dimension), with
+  the following symmetries.
+  
+    Minor symmetries:
+      C_ijkl = C_jikl = C_ijlk
+
+    Major symmetries:
+      C_ijkl = C_lkij
+
+  The minor symmetries are also reflected in the symmetric nature
+  of stress and strain tensors:
+    s_ij = s_ji
+    e_ij = e_ji
+
+  In general, there are 21 independent elastic constants in 3 dimension (6 in 2
+  dimensions). While systems with additional symmetries (e.g. isotropic,
+  orthotropic, etc.) can be expressed with fewer constants, we do not assume any
+  such additional symmetries.
+
+  At zero temperature, the response of every particle to a deformation can be 
+  calculated explicitly to linear order, enabling the exact calculation of 
+  the elastic modulus tensor without the need of any finite differences or any
+  approximations.
+
+  Args:
+    energy_fn: A function that computes the energy of the system. This
+      function must take as an argument `perturbation` which perturbes the
+      box shape. Any energy function constructed using `smap` or in `energy.py`
+      with a standard space will satisfy this property. 
+    tether_strength: scalar. Strength of the "tether" applied to each particle, 
+      which can be necessary to make the Hessian matrix non-singular. Solving
+      for the non-affine response of each particle requires that the Hessian is
+      positive definite. However, there can often be zero modes (eigenvectors of
+      the Hessian with zero eigenvalue) that do not couple to the boundary, and 
+      therefore do not affect the elastic constants despite the zero eigenvalue.
+      The most common example is the global translational modes. To solve for
+      the non-affine response, we consider the "tethered Hessian"
+        H + tether_strength * I, 
+      where I is the identity matrix and tether_strength is a small constant.
+    gradient_check: None or scalar. If not None, a check will be performed 
+      to guarantee that the maximum component of the gradient is less than
+      gradient_check. In other words, that
+        jnp.amax(jnp.abs(grad(energy_fn)(R, box=box))) < gradient_check == True
+      NOTE: JAX currently does not support proper runtime error handling.
+      Therefore, if this check fails, the calculation will return an array
+      of jnp.nan's. It is the users responsibility, if they want to use this
+      check, to then ensure that the returned array is not full of nans.
+    cg_tol: scalar. Tolorance used when solving for the non-affine response.
+    check_convergence: bool. If true, calculate_EMT will return a boolean
+      flag specifiying if the cg solve routine converged to the desired
+      tolorance. The default is False, but convergence checking is highly
+      recommended especially when using 32-bit precision data.
+  
+  Return: A function to calculate the elastic modulus tensor
+
+  """
+
+  def calculate_EMT(R: Array,
```

Comment:
Can the function name be lower case `calculate_emt`?

Reply:
done.